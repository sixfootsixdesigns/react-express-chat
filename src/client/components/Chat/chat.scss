html,
body,
#root {
  height: 100%;
  margin: 0;
  padding: 0;
}

.Chat {
  display: flex;
  flex-wrap: no-wrap;
  height: 100%;
  align-items: stretch;
}

.MessageWrapper {
  flex: 0 0 75%;
  max-width: 75%;
}

.MessageInner {
  display: flex;
  flex-direction: column;
  height: 100%;
}
.Messages {
  flex-grow: 1;
  overflow-y: auto;
  padding: 1rem;
  > div + div {
    margin-top: 0.5rem;
  }
}
.AddMessageWrapper {
  border-top: 1px solid gray;
  flex-grow: 0;
  padding: 0.5rem;
}
.AddMessage {
  display: flex;
  flex-wrap: nowrap;
  textarea {
    flex-grow: 1;
    max-width: 100%;
  }
}
.ActiveUsers {
  flex: 0 0 25%;
  max-width: 25%;
  border-left: 1px solid gray;
  padding: 1rem;
  h2 {
    margin: 0;
    padding: 0;
  }
  ul {
    list-style: none;
    margin: 0;
    padding: 0;
    li {
      padding: 0.5rem 0;
    }
  }
}

.JoinWrapper {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}
