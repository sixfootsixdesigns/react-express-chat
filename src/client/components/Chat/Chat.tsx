import React from 'react';
import { IMessage, IUser } from '../../../server/store';
import './chat.scss';
import { ActiveUsers } from '../ActiveUsers/ActiveUsers';
import { Messages } from '../Messages/Messages';
import { AddMessage } from '../AddMessage/AddMessage';
import { Join } from '../Join/Join';
import axios from 'axios';
import { socketIO } from '../../../server/app';

interface ChatState {
  messages?: IMessage[];
  user?: IUser;
  users?: IUser[];
  loadingMessages: boolean;
  loadingUsers: boolean;
}

interface ChatProps {
  socket: SocketIOClient.Socket;
}

export class Chat extends React.Component<ChatProps, ChatState> {
  private _isMounted: boolean = false;

  get isMounted(): boolean {
    return this._isMounted;
  }

  set isMounted(mountedStatus: boolean) {
    this._isMounted = mountedStatus;
  }

  public state = {
    messages: [],
    user: undefined,
    users: [],
    loadingMessages: false,
    loadingUsers: false
  };

  public handleBeforeUnload = () => {
    const { socket } = this.props;
    socket && socket.emit('disconnecting', this.state.user);
  };

  public componentDidMount() {
    this.isMounted = true;
    window.addEventListener('beforeunload', this.handleBeforeUnload);
  }

  public componentWillUnmount() {
    this.isMounted = false;
    window.removeEventListener('beforeunload', this.handleBeforeUnload);
  }

  addMessage = (message: string) => {
    axios
      .post('/add-message', {
        author: this.state.user,
        message
      })
      .catch(() => {
        console.log('error');
      });
  };

  addUser = (username: string) => {
    axios
      .post('/add-user', {
        username
      })
      .then(resp => {
        this.isMounted &&
          this.setState({
            user: resp.data
          });

        this.getInitialData();
      })
      .catch(() => {
        console.log('error');
      });
  };

  public render() {
    const { messages, users, user, loadingMessages, loadingUsers } = this.state;

    if (!user) {
      return <Join addUser={this.addUser} />;
    }

    return (
      <div className="Chat">
        <div className="MessageWrapper">
          <div className="MessageInner">
            <Messages loading={loadingMessages} messages={messages} />
            <AddMessage addMessage={this.addMessage} />
          </div>
        </div>
        <ActiveUsers loading={loadingUsers} currentUser={user} users={users} />
      </div>
    );
  }

  private getInitialData() {
    const { socket } = this.props;

    this.getActiveUsers();
    this.getMessages();

    if (socket) {
      socket.on('NewMessage', (messages: IMessage[]) => {
        if (!Array.isArray(messages)) {
          messages = [messages];
        }
        this.isMounted && this.setState({ messages });
      });

      socket.on('UserAdded', (users: IUser[]) => {
        if (!Array.isArray(users)) {
          users = [users];
        }
        this.isMounted && this.setState({ users });
      });

      socket.on('UserLeft', (users: IUser[]) => {
        if (!Array.isArray(users)) {
          users = [users];
        }
        this.isMounted && this.setState({ users });
      });
    }
  }

  private getActiveUsers() {
    axios
      .get('/users')
      .then(resp => {
        this.isMounted &&
          this.setState({
            users: resp.data
          });
      })
      .catch(er => {
        console.log('error');
      });
  }

  private getMessages() {
    axios
      .get('/messages')
      .then(resp => {
        this.isMounted &&
          this.setState({
            messages: resp.data
          });
      })
      .catch(er => {
        console.log('error');
      });
  }
}
