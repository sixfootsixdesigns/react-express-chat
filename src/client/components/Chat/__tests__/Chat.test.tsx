import * as React from 'react';
import {
  render,
  fireEvent,
  wait,
  getByText,
  getByPlaceholderText,
  cleanup
} from 'react-testing-library';
import { Chat } from '../Chat';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { SocketIO, Server } from 'mock-socket';

declare const window: any;

describe('Chat Component', () => {
  let axiosMock: MockAdapter;
  const fakeURL = 'ws://localhost:8080';
  const mockServer = new Server(fakeURL);
  window.io = SocketIO;
  const client = io(fakeURL);

  const joinChat = async (container: HTMLElement) => {
    const input = getByPlaceholderText(container, 'Enter your name');
    fireEvent.change(input, {
      target: { value: 'dave' }
    });

    const button = getByText(container, 'Join');
    fireEvent.click(button);

    await wait(() => {});
  };

  const postMessage = async (container: HTMLElement, message = 'hello') => {
    const input = getByPlaceholderText(container, 'Enter your message');
    fireEvent.change(input, {
      target: { value: message }
    });

    const button = getByText(container, 'Post');
    fireEvent.click(button);
    await wait(() => {});
  };

  beforeEach(() => {
    axiosMock = new MockAdapter(axios);
    axiosMock.onPost(`/add-user`).reply(201, { name: 'dave', id: '123' });
  });

  afterEach(() => {
    axiosMock.restore();
    cleanup();
  });

  it('should render with join screen', async () => {
    const { container } = render(<Chat socket={client} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render chat after joining', async () => {
    axiosMock.onGet(`/users`).reply(200, []);
    axiosMock.onGet(`/messages`).reply(200, []);

    const { container } = render(<Chat socket={client} />);

    await joinChat(container);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render existing users and messages once joined', async () => {
    axiosMock.onGet(`/users`).reply(200, [{ name: 'dave', id: '123' }]);
    axiosMock
      .onGet(`/messages`)
      .reply(200, [
        { author: { name: 'dave', id: '123' }, message: 'hi', id: 'zzz' }
      ]);

    const { container } = render(<Chat socket={client} />);

    await joinChat(container);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call add-message when posting a new message', async () => {
    axiosMock.onPost(`/add-message`).reply(201);
    axiosMock.onGet(`/users`).reply(200, []);
    axiosMock.onGet(`/messages`).reply(200, []);

    const { container } = render(<Chat socket={client} />);

    await joinChat(container);
    await postMessage(container);

    expect(axiosMock.history.post.length).toBe(2);
  });

  it('should add users if UserAdded is fired', async () => {
    axiosMock.onGet(`/users`).reply(200, []);
    axiosMock.onGet(`/messages`).reply(200, []);

    const { container } = render(<Chat socket={client} />);

    await joinChat(container);

    mockServer.emit('UserAdded', { name: 'bob', id: '10' });

    expect(container.querySelector('.ActiveUsers')).toMatchSnapshot();
  });

  it('should update users if UserLeft is fired', async () => {
    axiosMock
      .onGet(`/users`)
      .reply(200, [{ name: 'dave', id: '1' }, { name: 'tim', id: '2' }]);
    axiosMock.onGet(`/messages`).reply(200, []);

    const { container } = render(<Chat socket={client} />);

    await joinChat(container);

    expect(container.querySelector('.ActiveUsers')).toMatchSnapshot();

    mockServer.emit('UserLeft', { name: 'dave', id: '1' });

    expect(container.querySelector('.ActiveUsers')).toMatchSnapshot();
  });

  it('should add message if NewMessage is fired', async () => {
    axiosMock.onGet(`/users`).reply(200, []);
    axiosMock.onGet(`/messages`).reply(200, []);

    const { container } = render(<Chat socket={client} />);

    await joinChat(container);

    mockServer.emit('NewMessage', {
      author: { name: 'bob', id: '10' },
      message: 'henlo',
      id: 'ladfj'
    });

    expect(container.querySelector('.Messages')).toMatchSnapshot();
  });
});
