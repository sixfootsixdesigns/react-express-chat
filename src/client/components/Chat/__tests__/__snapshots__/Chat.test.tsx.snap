// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Component should add message if NewMessage is fired 1`] = `
<div
  class="Messages"
>
  <div>
    <strong>
      bob
    </strong>
    : 
    henlo
  </div>
</div>
`;

exports[`Chat Component should add users if UserAdded is fired 1`] = `
<div
  class="ActiveUsers"
>
  <h2>
    Users
  </h2>
  <ul>
    <li>
      bob
    </li>
  </ul>
</div>
`;

exports[`Chat Component should render chat after joining 1`] = `
<div
  class="Chat"
>
  <div
    class="MessageWrapper"
  >
    <div
      class="MessageInner"
    >
      <div
        class="Messages"
      >
        <div>
          No Messages Yet.
        </div>
      </div>
      <div
        class="AddMessageWrapper"
      >
        <div
          class="AddMessage"
        >
          <textarea
            placeholder="Enter your message"
          />
          <button>
            Post
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ActiveUsers"
  >
    <h2>
      Users
    </h2>
    <p>
      No Active Users
    </p>
  </div>
</div>
`;

exports[`Chat Component should render existing users and messages once joined 1`] = `
<div
  class="Chat"
>
  <div
    class="MessageWrapper"
  >
    <div
      class="MessageInner"
    >
      <div
        class="Messages"
      >
        <div>
          <strong>
            dave
          </strong>
          : 
          hi
        </div>
      </div>
      <div
        class="AddMessageWrapper"
      >
        <div
          class="AddMessage"
        >
          <textarea
            placeholder="Enter your message"
          />
          <button>
            Post
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ActiveUsers"
  >
    <h2>
      Users
    </h2>
    <ul>
      <li>
        ** 
        dave
      </li>
    </ul>
  </div>
</div>
`;

exports[`Chat Component should render with join screen 1`] = `
<div
  class="JoinWrapper"
>
  <div
    class="Join"
  >
    <input
      placeholder="Enter your name"
      required=""
      type="text"
      value=""
    />
    <button>
      Join
    </button>
  </div>
</div>
`;

exports[`Chat Component should update users if UserLeft is fired 1`] = `
<div
  class="ActiveUsers"
>
  <h2>
    Users
  </h2>
  <ul>
    <li>
      dave
    </li>
    <li>
      tim
    </li>
  </ul>
</div>
`;

exports[`Chat Component should update users if UserLeft is fired 2`] = `
<div
  class="ActiveUsers"
>
  <h2>
    Users
  </h2>
  <ul>
    <li>
      dave
    </li>
  </ul>
</div>
`;
